Added `disallow-multiple-spaces` rule. Fixes #435

twolfson opened the pull request:
This PR adds a rule for disallowing multiple spaces between any tokens (e.g. `var` and `x`). I decided to add to `token-assert` for reusability purposes. In this PR:

- Added tests for `disallow-multiple-spaces`
- Added rule for `disallow-multiple-spaces`
- Added `maximumSpaces` option for `TokenAssert.whitespaceBetween`

**Missing:**

- Negation rule for `disallow-multiple-spaces` (e.g. `require-multiple-spaces`)
    - I started building this but realized that it was unlikely to be used and none of the other whitespace `disallow` rules had negations
        - https://github.com/twolfson/node-jscs/compare/496c8911bb1f6d9ccadd98e201558e4b6c344ae4

twolfson commented:
Ah, I forgot to test multiple lines. I will add that shortly.

zxqfox commented on test/rules/disallow-multiple-spaces.js:
```diff
@@ -0,0 +1,38 @@
+
+    it('should not report multiple spaces in strings/regular expressions', function() {
+        assert(checker.checkString('"hello  world";').isEmpty());
+        assert(checker.checkString('/hello  world/').isEmpty());
```
Can you add some tests with comments? Like:
```
var X =  /** @lends X */  {  /** @type {String} */  name: "some" };
```

twolfson replied:
Sure thing, more tests coming right up.

zxqfox commented on lib/rules/disallow-multiple-spaces.js:
```diff
@@ -0,0 +1,65 @@
+    },
+
+    check: function(file, errors) {
+        file.getTokens().forEach(function(token, index, tokens) {
```
I'm not sure but `getTokens` returns token list without comments and tests suggested by me will fail.
https://github.com/jscs-dev/node-jscs/blob/master/lib/js-file.js#L492

I suggest to use here `while (token = file.nextToken({includeComments: true}) { ...` or something similar to iterate through tokens and comments.

zxqfox replied:
@mikesherov @mdevils What you think about integrating this into `JSFile.interateTokens*` methods?

twolfson replied:
Ah, didn't realize we were missing those tokens. Good catch. Do you think there would be a reuse value in an `iterateTokens` with options for including comment tokens?

twolfson commented:
@zxqfox I have added in the comment tests and fix for iterating with comments. The `file.getNextToken` method requires inputting a seed token. Unfortunately, there is no way to get the first token of a file without accessing `_tokens`. As a result, I I chose to replace our `getTokens` with a direct iterator on `_tokens`.

zxqfox commented on test/rules/disallow-multiple-spaces.js:
```diff
@@ -0,0 +1,56 @@
+
+    it('should report multiple spaces between comments', function() {
+        assert(checker.checkString('var X = {  /** @type {String} */ name: "some" };').getErrorCount() === 1);
+        assert(checker.checkString('var x = "oops";  // Multiple spaces before comment').getErrorCount() === 1);
```
I think this case should be an exception. Maybe we can make an option?
```
var CONST1 = "one"; // makes me sad
var C2 = "1";       // makes me happy
```

twolfson replied:
Yep, definitely. I will add a new option:

```js
disallowMultipleSpace: {
  allowEOLComments: true
}
```

zxqfox commented on lib/token-assert.js:
```diff
@@ -38,6 +38,21 @@ TokenAssert.prototype.whitespaceBetween = function(options) {
                 column: token.loc.end.column
             });
         }
+    } else if (options.hasOwnProperty('maximumSpaces')) {
```
This is very close with https://github.com/jscs-dev/node-jscs/pull/1023/files#diff-526f88acb4a66447fc67eaf20ca6acadR104

We need to decide what naming we should use.
/cc @markelog @mdevils @mikesherov

twolfson replied:
We should probably clarify that those are about lines. It will be confusing to have them and `maximumSpaces`. We can rename to `atMostSpaces` for consistency.

twolfson replied:
Ah, nvm. Fell into the same trap. That is a different function entirely. I will still rename to `atLeastSpaces` for semi-consistency but whitespace isn't necessarily spaces only.

zxqfox replied:
`atMost` option is quietly fine if method will be about spaces (if we decide to use `atMost` for lines). I just wanted some consistency in options naming.

twolfson replied:
There is a possible trap if we ever start to use `tabs` inside of `whitespaceBetween` but I will move to `atMost` for now. I am going to add a comment on the other PR about renaming `exactly` to `lines` for consistency with `whitespaceBetween`.

zxqfox replied:
Good idea ;-)

zxqfox commented:
@twolfson Thanks for contribution! Great work here. Here we have a conflict with another PR so I suggest to wait someone else to discuss with.
Personally, I like atLeast/atMost (just because it's shorter). ;-)

twolfson commented:
I have added the case for EOL comments and renamed to `atMost`. After the rename, I did a `git grep` and saw that we are using `maximumLineLength` as a rule and not using `atMost` anywhere else. That might have a factor on naming.

mrjoelkemp commented:
@mikesherov cool with this?

@twolfson Mind rebasing?

mikesherov commented:
I'm not really a fan of this rule because I think it's solved by adding `atLeast`, `atMost`, and `exactly` to the spacing assertions and passing those down through the rule config themselves. This is a fairly decent stop gap measure but I predict will be rife with requests for people asking for leniency to this rule in certain cases, which highlights the problem of it being too overarching.

vkramskikh commented:
I'd really love to see this in master. Double spaces is the most frequent error in my project which is still not covered by JSCS.

twolfson commented:
@mikesherov I am not sure I follow. Do you mean adding `atLeast`/`atMost`/`exactly` support to every `requireSpace*` rule?

twolfson commented:
@mrjoelkemp I will take a shot at rebasing by the end of this weekend. It looks like there are some merge conflicts as well =/

mikesherov commented:
I'm fine merging it in, I was just saying that the customizable route will be not to address it by making this rule have tons of options, but rather by enabling more fine grained control of the other options.

twolfson commented:
I have merged/squashed on top of the latest `master`. Notable changes:

- Relocated test files/changes for consistency
- Renamed `configure` parameter to `options` for consistency
- Moved to `file.getTokens` over `file._tokens` to stop relying on internals

mrjoelkemp commented:
Nice work here @twolfson. Always happy to see a super old issue closed. Thanks for contributing.
